os/fs/driver/mtd/smart.c

```diff
@@ -5408,16 +5425,19 @@ static int smart_journal_scan(FAR struct smart_struct_s *dev, bool print_dump)
 static int smart_journal_recovery(FAR struct smart_struct_s *dev, journal_log_t *log)
 {
 	int ret;
-	crc_t crc; //journal crc
 	int type;
 	
 	fvdbg("Printf recovery data journal_seq : %d!!\n", dev->journal_seq);
 	smart_journal_print_log(dev, log);
 
-	crc = smart_calc_journal_crc(log);
-
-	if (crc != UINT8TOUINT16(log->crc16)) {
-		fdbg("Invalid CRC %d calculated crc : %d written crc : %d\n", crc, log->crc16);
+	/* Recovery step is Check crc -> Check something more based on type -> checkout -> verify based on type */
+	if (smart_validate_journal_crc(log) != OK) {
+		fdbg("Invalid CRC %d calculated crc : %d written crc : %d\n", smart_calc_journal_crc(log), log->crc16);
+		/* Invalid one, so make it checkout */
+		ret = smart_journal_checkout(dev, log, address);
+		if (ret != OK) {
+			return ret;
```

Comment:
This will never return OK.

Reply:
No if checkout finished properly, ret is OK but crc is not valid so finally it return -EINVAL

Reply:
If the ret value is not OK, the ret value is always -EINVAL.

Reply:
It should return -EINVAL because it is invalid journal, it means that power failed during writing journal.
So it should checkout to disable target journal, but inside of smart_journal_checkout I mean old one, there were crc check logic so it alwasys return -EIO.
So I divided smart_journal_checkout.